### ISSUE TITLE
Bug? BepInEx does not *seem* to pass Startup Args to Valheim [Windows/Steam]

### ISSUE THREAD

**Issue by Formula350** [BepInEx/BepInEx#223]
**Describe the bug**
From what I can tell based on my limited Not a Coder experience, is that my Unity startup argument of `-force-d3d12` is not being passed to Valheim after BepInEx intercepts everything and starts loading Mods (plugins)... Mind you, I'm basing this on a a couple things, but can't actually confirm it to be the case.

For starters is that after installing Bep I noticed that there was definitely a vid restart occurring during what seems to be 2 points. First one happens after the intro logos (before Main Menu displays), and the second seems to happen after selecting the World to load (ie beginning your gameplay). The first one seems to be the main issue as that's when I get the impression the argument isn't being passed. (similar to #186 maybe?)

I'm also wondering if boot.cfg is discarded too, but I really don't know... (Not that I know for certain that it's not using DX12 lol; will include that in Further Context section)

**To Reproduce**
In Steam client: View properties on the game (Valheim in my case) and in the window that displays, insert `-force-d3d12`
Accept and close the window.
Launch the game.
*NOTE: You may need to launch and configure the game to run in Fullscreen mode, exit and relaunch, since pretty sure Valheim defaults to Windowed mode.*

**Expected behavior**
There should be no discernible screen flashing/flickering, indicating a change in graphics behavior (commonly experienced when changing refreshrate in a game which is running in fullscreen), yet with Bep installed there definitely is. 
*[Whether relevant, previously I only noticed this behavior on launching when forcing Vulkan, prior to installing Bep]*

**Screenshots and logs**
A snippet from Unity's log in AppData
```[Subsystems] Discovering subsystems at path D:/!GAMES!/SteamLibrary/steamapps/common/Valheim/valheim_Data/UnitySubsystems
Forcing GfxDevice: Direct3D 12                    <-------------
GfxDevice: creating device client; threaded=1
d3d12: loaded!                    <-------------
Direct3D:
    Version:         Direct3D 12 [level 12.0]                    <-------------
    Renderer:        AMD Radeon (TM) R9 390 Series (ID=0x67b1)
    Vendor:          
    VRAM:            8168 MB
    App VRAM Budget: 6943 MB
    Driver:          25.20.15002.58
Begin MonoManager ReloadAssembly
- Completed reload, in  0.052 seconds
D3D11 device created for Microsoft Media Foundation video decoding.     <<==========
<RI> Initializing input.
<RI> Input initialized.
<RI> Initialized touch support.
UnloadTime: 0.638698 ms
[Message:   BepInEx] BepInEx 5.4.9.0 - valheim
[Info   :   BepInEx] Running under Unity v2019.4.20.7197120```

This is approximately the time when the Company Logo intros end and the flicker occurs and then Busy cursor starts (which I attribute it being 'busy' due to Bep processing mods):
```[Message:   BepInEx] Chainloader startup complete
[Info   : Unity Log] 04/07/2021 20:33:46: Starting to load scene:start
```
Once that pause is finished, then the "Valheim logo" displays, which seems to coincide with this entry (30sec later)...
```[Info   : Unity Log] 04/07/2021 20:34:06: Valheim version:0.148.7```

Then the Main Menu shows up a moment later...
```[Info   : Unity Log] 04/07/2021 20:34:11: UI Group status changed Menu = True```
**Desktop (please complete the following information):**
 - Valheim v0.148.7
 - [BepInEx Pack for Valheim](https://valheim.thunderstore.io/package/denikson/BepInExPack_Valheim/) v5.4.900

**Additional context**
I know that it is at least _starting_ in DX12 (beyond the log bits), because pre-Bep usage, I was testing the different APIs to see which worked best on my system. To which by default Valheim uses DX11, with a startup variable from Steam prompting to use Vulkan (though this only occurred once and apparently it is bugged, as subsequent launches are reported by some to not be launching in Vulkan?). First time I launched Valheim I selected Vulkan and for whatever reason decided to test the others, as I knew Unity easily let you. After forcing DX11, 12 and Vulkan, I concluded that for my computer, DX12 yielded 3-5FPS higher, and so I added the argument to the Steam properties.

Now back to my point, starting with DX12 forced, during the intro clips, the Steam-overlay FPS counter skyrockets to >3200, whereas in Vulkan and DX11 it remains clamped to VSync refresh since I have VSync enabled in the game (65FPS due to 65Hz in my case). This insane FPS still occurs with Bep installed, which is why I know that it does at least startup in DX12, but, that's pre screen flickering/busy wheel.

As for my Boot.cfg suspicion, I'm not super sure that's happening, however, I noticed in the log something that had me wonder, and also sorta made me feel it might be BepInEx behind it.
In my boot.cfg file I have it set with `gfx-disable-mt-rendering=1`, although Valheim by default is `0`. 
Yet in the AppData Player.log file it displays this, which seems to coincide with Valheim logo showing up, 5sec before main menu displays:
```04/07/2021 20:34:06: Render threading mode:MultiThreaded```

But, admittedly, that could be unrelated to the `boot.cfg` line and some other internal and similarly named feature, **OR** it could be actually to do with DX12's multithreaded nature and indicating that I'm wrong on this entire bug report since it would still be in DX12 mode (despite this coming after the "DX11" log entry).

Yea ok, this came out really rambly, I apologize! :( Will try to clarify any points that need to be...

**Comment by Formula350** [comment]
Crap, Markdown error on my part, and the Edit button isn't working for me... 
That  \`\`\` after the first log chunk needed to be on its own line.

**Comment by ghorsington** [comment]
Greetings!

Thank you for a detailed report. First of all, a clarification: Valheim itself does not handle `-force-d3d12`. Instead, it is handled by the underlying Unity engine which has [a set of different command launch arguments](https://docs.unity3d.com/Manual/CommandLineArguments.html).

Next, I unfortunately cannot replicate your problem.

Launching the game with `-force-d3d12` yields correct log:

```
Initialize engine version: 2019.4.20f1 (6dd1c08eedfa)
Forcing GfxDevice: Direct3D 12
GfxDevice: creating device client; threaded=1
d3d12: loaded!
```

and the game boots fine without flickers nor flashes (mind you, I'm using NVIDIA GeForce RTX 2060 SUPER for testing). Moreover, looking at the logs there is no indication that the game would've been restarted or if the graphics device was reinitialized. 

As such, the command line argument does correctly get passed to Unity. BepInEx itself does not modify Unity in any way, it just injects a game object at the game start. The engine itself is not modified in any way. Unfortunately you would have to give this a bit more testing. I don't see how BepInEx would affect how the graphics device is initialized on Unity side given that BepInEx has nothing to do with that.

**Comment by Formula350** [comment]
I can completely accept that I could be, or flat out am, wrong here! 
In general the flicker doesn't always happen, but I think some of the times when I've noticed it, it may be due to the strange fact that\* on my system when I launch Valheim(Unity) with BepInEx installed, it's actually forced into the background instead, and the Bep console is what almost always is left in focus. So on occasion it may be that the flicker is due to the game threads being slightly lower priority. Which I think is rather likely since those extreme 3000FPS instances are only when I've clicked the game into focus. Otherwise, if it's kept in the background the FPS are clamped to VSync refresh (65FPS in my case). 
*\*[I noticed in the "for Valheim" package's Readme, a change made for this variant has it set to always show the console, and I tried the supplied arg `--enable-console false` which allowed the game to load ""normally"".]*

Either way, the main basis of why I thought it was failing to remain in DX12 was due to this line (which I marked with the <<==== originally)
`D3D`**`11`**`device created for Microsoft Media Foundation video decoding.`
Granted, I know that DirectX is an ongoing and continuous upgrade of previous versions, and DX11 was very much DX10 with only a few extra defining features. As such, I suppose a string could have been copy-pasted (either on Unity's side, or even Microsoft's) from the DX11-specific strings table, into the DX12-specific strings, and they just forgot to change the 11 into a 12.

This is what is displayed instead, when running Vulkan:
`Microsoft Media Foundation video decoding to texture disabled: graphics device is Vulkan, only Direct3D 11 and Direct3D 12 (only on desktop) are supported.`


That being said, if you're confident that nothing here is indicating what I was thinking, I'm content with that verdict and then this can be closed! :)

**Comment by ghorsington** [comment]
If I understood correctly, disabling the console fixes the issues? In that case it's very likely a wontfix for now since we don't really have proper control over how Windows spawns and handles the attached console. There are plans to make an alternative terminal emulator that would allow for better multiplatform support and general better control over console properties.

If it helps, you can disable BepInEx console via `BepInEx.cfg` instead of a command line arguments.  
I'll close the issue for now as it seems like a combination of wontfix and lacking info as I am unable to replicate this at all.

If there is more concrete info on this, do feel free to comment on the ticket. I'll try to replicate it on my side more and reopen a ticket if I do manage to do so.
